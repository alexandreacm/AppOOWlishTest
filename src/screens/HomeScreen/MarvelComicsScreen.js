import React, { useCallback, memo, useEffect, useState, useRef } from 'react';

import { FlatList } from 'react-native';

import styled, { css } from 'styled-components/native';
import { useSelector, useDispatch } from 'react-redux';
import { useNavigation } from '@react-navigation/native';

import TextInput from '@/components/TextInput';
import Card from '@/components/Card';

import { LOADING_MARVEL_CHARACTERS } from '@/store/slices/marvelComicsSlice';

const MarvelComics = () => {
  const { navigate } = useNavigation();
  const dispatch = useDispatch();
  const [nameCharacter, setNameCharacter] = useState('');
  const passwordRef = useRef(null);

  const { isLoading, listCharactersData, hasError, errorMessage } = useSelector(
    ({ comics }) => comics
  );

  const handleTitleSubmit = () => passwordRef.current?.focus();

  const onHandleNavigate = useCallback(
    id => {
      navigate('details-comics', { id });
    },
    [navigate]
  );

  const renderItem = ({ item: { id, thumbnail, name, description } }) => (
    <StyledCardCharacter
      key={id}
      id={id}
      image={{
        uri: `${thumbnail.path}.${thumbnail.extension}`
      }}
      onPress={() => onHandleNavigate(id)}
      name={name}
      description={description || 'Sem descrição do herói'}
    />
  );

  useEffect(() => {
    dispatch(LOADING_MARVEL_CHARACTERS({ nameCharacter }));
  }, [dispatch, nameCharacter]);

  return (
    <>
      <StyledHeader>
        <TextInput
          onSubmitEditing={handleTitleSubmit}
          width='100%'
          label='Pesquisar Heróis'
          value={nameCharacter}
          onChangeText={setNameCharacter}
          marginBottom={8}
          hasError={hasError}
        />
        {errorMessage && (
          <StyledViewLabe>
            <StyledLabel>{errorMessage}</StyledLabel>
          </StyledViewLabe>
        )}
      </StyledHeader>
      <StyledContainer>
        {isLoading ? (
          <StyledLoading />
        ) : (
          <FlatList
            data={listCharactersData}
            keyExtractor={item => String(item.id)}
            renderItem={renderItem}
          />
        )}
      </StyledContainer>
    </>
  );
};

const StyledContainer = styled.View`
  flex: 1;
  background-color: #fff;
  margin-left: 10px;
  margin-right: 10px;
  margin-top: 20px;
`;

const StyledHeader = styled.View`
  width: 100%;
  justify-content: center;
  align-items: center;
  margin-top: 10px;
  padding: 10px;
`;

const StyledLoading = styled.ActivityIndicator.attrs(
  ({ theme: { colors } }) => ({
    color: colors.SECUNDARY
  })
)``;

const StyledLabel = styled.Text`
  width: 100%;
  font-weight: 400;
  font-size: 14;
  margin-bottom: 8;
  ${({ theme: { colors } }) => css`
    color: ${colors.COLOR_WHITE}};
  `};
`;
const StyledViewLabe = styled.View`
  width: 100%;
  padding: 10px;
  align-items: center;
  ${({ theme: { colors } }) => css`
    background-color: ${colors.DANGER}};
  `};
`;

const StyledCardCharacter = styled(Card)`
  margin-bottom: 8px;
`;

export default memo(MarvelComics);
