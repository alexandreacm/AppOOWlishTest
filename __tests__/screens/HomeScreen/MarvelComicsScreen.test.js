import 'react-native';
import React from 'react';
import {
  matchSnapshotWithTheme,
  renderWithTheme
} from '@/helpers/functions/testing';

import MarvelComicsScreen from '@/screens/HomeScreen/MarvelComicsScreen';

describe('tests in MarvelComicsScreen Screen', () => {
  test('matches snapshot', () => {
    matchSnapshotWithTheme(<MarvelComicsScreen />);
  });

  test('renders correctly', () => {
    renderWithTheme(<MarvelComicsScreen />);
  });
});
