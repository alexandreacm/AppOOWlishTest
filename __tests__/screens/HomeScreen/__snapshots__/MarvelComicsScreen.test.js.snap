// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in MarvelComicsScreen Screen matches snapshot 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginTop": 10,
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
          "width": "100%",
        },
      ]
    }
  >
    <View
      borderColor="#878787"
      marginBottom={8}
      marginLeft={0}
      marginRight={0}
      marginTop={0}
      paddingBottom={0}
      paddingLeft={16}
      paddingRight={16}
      paddingTop={0}
      style={
        Array [
          Object {
            "borderColor": "#878787",
            "borderRadius": 8,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 56,
            "justifyContent": "space-between",
            "marginBottom": 8,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "paddingBottom": 0,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 0,
            "width": "100%",
          },
        ]
      }
      width="100%"
    >
      <Text
        backgroundColor="#FFF"
        color="#878787"
        style={
          Object {
            "backgroundColor": "#FFF",
            "color": "#878787",
            "fontSize": 16,
            "left": 16,
            "letterSpacing": 0.15,
            "lineHeight": 24,
            "paddingLeft": 4,
            "paddingRight": 4,
            "position": "absolute",
            "top": 16,
          }
        }
      >
        Pesquisar Her√≥is
      </Text>
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCompleteType="off"
        autoFocus={false}
        blurOnSubmit={true}
        editable={true}
        keyboardType="default"
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        rejectResponderTermination={true}
        returnKeyType="done"
        secureTextEntry={false}
        style={
          Array [
            Object {
              "color": "#878787",
              "fontSize": 16,
              "width": "90%",
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Array [
            Object {
              "justifyContent": "center",
            },
          ]
        }
      />
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "marginLeft": 10,
          "marginRight": 10,
          "marginTop": 20,
        },
      ]
    }
  >
    <RCTScrollView
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View />
    </RCTScrollView>
  </View>,
]
`;
